### PR TITLE
ReplicationTracker.markAllocationIdAsInSync may hang if allocation is cancelled

### PR DESCRIPTION
**Pull request by bleskes**
At the end of recovery, we mark the recovering shard as "in sync" on the primary. From this point on the primary will treat any replication failure on it as critical and will reach out to the master to fail the shard. To do so, we wait for the local checkpoint of the recovered shard to be above the global checkpoint (in order to maintain global checkpoint invariant).

If the master decides to cancel the allocation of the recovering shard while we wait, the method can currently hang and fail to return. It will also ignore the interrupts that are triggered by the cancelled recovery due to the primary closing. 

Note that this is crucial as this method is called while holding a primary permit. Since the method never comes back, the permit is never released. The unreleased permit will then block any primary relocation *and* while the primary is trying to relocate all indexing will be blocked for 30m as it waits to acquire the missing permit.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java [hunk 1]
```diff
@@ -339,6 +339,11 @@ private boolean invariant() {
                 "shard copy " + entry.getKey() + " is in-sync but not tracked";
         }
 
+        // all pending in sync shards are tracked
+        for (String aId: pendingInSync) {
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: Nit: `aId:` -> `aId : `

**Review by jasontedor — APPROVED** [review]
I left an annoying nit but you made the typo. 😇 

LGTM.

**Review by ywelsch — APPROVED** [review]
LGTM. This looks to be the cause for the test failures seen here: https://github.com/elastic/elasticsearch/issues/29161#issuecomment-384981315

**Comment by ywelsch** [comment]
As it's a blocker, can you label it 6.3.0?

**Comment by bleskes** [comment]
> As it's a blocker, can you label it 6.3.0?

I agree it's confusing. I was following the standard labelling of 6.3.1 + relabel on respin (once we know it ships). I'll change the label.

**Comment by bleskes** [comment]
run gradle build tests

**Comment by bleskes** [comment]
Thanks @jasontedor @ywelsch

**Pull request merged**